src/sage/combinat/bijectionist.py

```diff
@@ -0,0 +1,3217 @@
+    sage: value_restrictions = [(Permutation([1, 2]), [1]), (Permutation([3, 2, 1]), [2, 3, 4])]
+    sage: bij = Bijectionist(A, B, tau, alpha_beta=alpha_beta, value_restrictions=value_restrictions)
+    sage: for sol in sorted(list(bij.solutions_iterator()), key=lambda d: tuple(sorted(d.items()))):
+    ....:     print(sol)
+    {[]: 0, [1]: 1, [1, 2]: 1, [2, 1]: 2, [1, 2, 3]: 1, [1, 3, 2]: 2, [2, 1, 3]: 2, [2, 3, 1]: 2, [3, 1, 2]: 2, [3, 2, 1]: 3}
+    ...
+    {[]: 0, [1]: 1, [1, 2]: 1, [2, 1]: 2, [1, 2, 3]: 3, [1, 3, 2]: 2, [2, 1, 3]: 2, [2, 3, 1]: 2, [3, 1, 2]: 1, [3, 2, 1]: 2}
+
+    sage: A = B = [permutation for n in range(4) for permutation in Permutations(n)]
+    sage: tau = Permutation.longest_increasing_subsequence_length
+    sage: bij = Bijectionist(A, B, tau, value_restrictions=((Permutation([1, 2]), [4, 5]),))
+    Traceback (most recent call last):
+    ...
+    ValueError: No possible values found for singleton block [[1, 2]]
+
+"""
+# ****************************************************************************
+#       Copyright (C) 2020 Martin Rubey <martin.rubey at tuwien.ac.at>
+#                          Stephan Pfannerer
+#                          Tobias Kietreiber
+#                          Alexander Grosz
+#
+#  Distributed under the terms of the GNU General Public License (GPL)
+#
+#    This code is distributed in the hope that it will be useful,
+#    but WITHOUT ANY WARRANTY; without even the implied warranty of
+#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+#    General Public License for more details.
+#
+#  The full text of the GPL is available at:
+#
+#                  https://www.gnu.org/licenses/
+# ***************************************************************************
+import itertools
+from collections import namedtuple, defaultdict
+from sage.numerical.mip import MixedIntegerLinearProgram, MIPSolverException
+from sage.rings.integer_ring import ZZ
+from sage.combinat.set_partition import SetPartition, SetPartitions
+from sage.sets.disjoint_set import DisjointSet
+from sage.structure.sage_object import SageObject
+from copy import copy
+from sage.misc.verbose import get_verbose
+
+
+class Bijectionist(SageObject):
+    r"""
+    A toolbox to list all possible bijections between two finite sets
+    under various constraints.
+
+    INPUT:
+
+    - ``A``, ``B`` -- sets of equal size, given as a list
+
+    - ``tau`` -- (optional) a function from ``B`` to ``Z``, in case of
+      ``None``, the identity map ``lambda x: x`` is used
+
+    - ``alpha_beta`` -- (optional) a list of pairs of statistics ``alpha`` from
+      ``A`` to ``W`` and ``beta`` from ``B`` to ``W``
+
+    - ``P`` -- (optional) a partition of ``A``
+
+    - ``pi_rho`` -- (optional) a list of triples ``(k, pi, rho)``, where
+
+        - ``pi`` -- a ``k``-ary operation composing objects in ``A`` and
+
+        - ``rho`` -- a ``k``-ary function composing statistic values in `Z`
+
```

Comment:
```suggestion
    - ``pi_rho`` -- (optional) a list of triples ``(k, pi, rho)``, where

      * ``pi`` -- a ``k``-ary operation composing objects in ``A`` and
      * ``rho`` -- a ``k``-ary function composing statistic values in ``Z``
```
The biggest thing is the indentation on the sublist.